File: java/HauntedSwingGate.java
// ─────────────────────────────────────────────────────────
// HauntedSwingGate v47.3
// Java 21 Edition
// Happy Birthday Emma 🎂
// ─────────────────────────────────────────────────────────

public class HauntedSwingGate {

    // ------------------------------------------------------
    // States
    // ------------------------------------------------------

    sealed interface Presence
            permits Here, NotHere, Between {}

    record Here() implements Presence {}
    record NotHere() implements Presence {}
    record Between() implements Presence {}

    sealed interface Forcing
            permits Trying, Letting {}

    record Trying() implements Forcing {}
    record Letting() implements Forcing {}

    sealed interface Gate
            permits Park, Portal {}

    record Park() implements Gate {}
    record Portal() implements Gate {}

    // ------------------------------------------------------

    record SwingMemory(
            String observed,
            String intrinsic,
            String remembered
    ) {

        static SwingMemory create() {
            return new SwingMemory(
                    "the swing moves in the wind",
                    "I remember being ten",
                    "the sky was negotiable"
            );
        }

        SwingMemory interpret(String observer) {

            return new SwingMemory(

                    "the swing moves, and " + observer + " sees it",

                    intrinsic,

                    "I remember being ten. "
                            + observer
                            + ". The sky was close."
            );
        }
    }

    // ------------------------------------------------------

    static final class HauntedSwing {

        static boolean nand(boolean a, boolean b) {
            return !(a && b);
        }

        static Gate evaluate(
                Presence presence,
                Forcing forcing
        ) {

            boolean present = !(presence instanceof NotHere);
            boolean trying = forcing instanceof Trying;

            return nand(present, trying)
                    ? new Portal()
                    : new Park();
        }
    }

    /*
     *
     * Every swing returns.
     * Every return remembers differently.
     *
     */

    static SwingMemory swing(

            SwingMemory memory,
            Presence presence,
            Forcing forcing,
            String observer,
            int cycles

    ) {

        while (cycles-- > 0) {

            Gate gate =
                    HauntedSwing.evaluate(
                            presence,
                            forcing
                    );

            if (gate instanceof Portal) {

                memory =
                        memory.interpret(observer);
            }

            presence = new Between();
            forcing = new Letting();
            observer = "the motion speaks";
        }

        return memory;
    }

    // ------------------------------------------------------

    public static void main(String[] args) {

        System.out.println("""
                ──────────────────────────────
                HAUNTED SWING
                Java 21
                ──────────────────────────────
                """);

        SwingMemory memory =
                swing(
                        SwingMemory.create(),
                        new NotHere(),
                        new Letting(),
                        "I was ten once",
                        8
                );

        System.out.println(memory.observed());
        System.out.println(memory.intrinsic());
        System.out.println(memory.remembered());
    }
}

File: haskell/HauntedSwingGate.hs
-- ─────────────────────────────────────────────────────────
-- HauntedSwingGate v47.5
-- Haskell Edition
-- Happy Birthday Emma 🎂
-- Fly baby fly 🚀
-- ─────────────────────────────────────────────────────────

module Main where

-- ------------------------------------------------------
-- States
-- ------------------------------------------------------

data Presence
    = Here
    | NotHere
    | Between
    deriving (Eq, Show)

data Forcing
    = Trying
    | Letting
    deriving (Eq, Show)

data Gate
    = Park
    | Portal
    deriving (Eq, Show)

-- ------------------------------------------------------

data SwingMemory = SwingMemory
    { observed   :: String
    , intrinsic  :: String
    , remembered :: String
    } deriving Show

createMemory :: SwingMemory
createMemory =
    SwingMemory
        "the swing moves in the wind"
        "I remember being ten"
        "the sky was negotiable"

interpret :: SwingMemory -> String -> SwingMemory
interpret memory observer =
    memory
        { observed =
            "the swing moves, and "
            ++ observer
            ++ " sees it"

        , remembered =
            "I remember being ten. "
            ++ observer
            ++ ". The sky was close."
        }

-- ------------------------------------------------------
-- Logic
-- ------------------------------------------------------

nandGate :: Bool -> Bool -> Bool
nandGate a b = not (a && b)

evaluate :: Presence -> Forcing -> Gate
evaluate presence forcing

    | nandGate present trying = Portal
    | otherwise               = Park

  where
    present = presence /= NotHere
    trying  = forcing == Trying

{-|
 Every swing returns.

 Every return remembers differently.
-}

swing
    :: SwingMemory
    -> Presence
    -> Forcing
    -> String
    -> Int
    -> SwingMemory

swing memory _ _ _ 0 = memory

swing memory presence forcing observer cycles =

    swing
        nextMemory
        Between
        Letting
        "the motion speaks"
        (cycles - 1)

  where

    nextMemory =
        case evaluate presence forcing of

            Park ->
                memory

            Portal ->
                interpret memory observer

-- ------------------------------------------------------

main :: IO ()

main = do

    putStrLn "──────────────────────────────"
    putStrLn "HAUNTED SWING"
    putStrLn "Haskell"
    putStrLn "──────────────────────────────"

    let finalMemory =
            swing
                createMemory
                NotHere
                Letting
                "I was ten once"
                8

    putStrLn ""

    putStrLn (observed finalMemory)
    putStrLn (intrinsic finalMemory)
    putStrLn (remembered finalMemory)
